File: src/main/java/io/wisetime/idea/branch/BranchTitleInfoProvider.java
/*
 * Copyright (c) 2023 Practice Insight Pty Ltd. All Rights Reserved.
 */

package io.wisetime.idea.branch;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.wm.impl.TitleInfoProvider;
import java.util.Optional;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import org.jetbrains.annotations.NotNull;

public class BranchTitleInfoProvider implements TitleInfoProvider {

  @NotNull
  @Override
  public String getBorderlessPrefix() {
    return " ";
  }

  @NotNull
  @Override
  public String getBorderlessSuffix() {
    return "";
  }

  @Override
  public void addUpdateListener(@NotNull Project project, @NotNull Disposable disposable,
      @NotNull Function1<? super TitleInfoProvider, Unit> callback) {
  }

  @NotNull
  @Override
  public String getValue(@NotNull Project project) {
    return Optional.ofNullable(project.getService(BranchHelper.class).getCurrentBranchName())
        .orElse("");
  }

  @Override
  public boolean isActive(@NotNull Project project) {
    return true;
  }

}

File: src/main/kotlin/io/wisetime/idea/branch/BranchTitleInfoProvider.kt
package io.wisetime.idea.branch

import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.openapi.project.Project
import com.intellij.openapi.wm.impl.TitleInfoProvider

class BranchTitleInfoProvider : TitleInfoProvider {

  override val borderlessPrefix: String
    get() = " "

  override val borderlessSuffix: String
    get() = ""

  override fun addUpdateListener(
    project: Project,
    disp: Disposable,
    value: (provider: TitleInfoProvider) -> Unit
  ) {
  }

  override fun getValue(project: Project): String {
    val service = project.getService(BranchHelper::class.java)
    if (service == null) {
      thisLogger().warn("Failed to get BranchHelper service")
      return ""
    }
    return service.currentBranchName ?: ""
  }

  override fun isActive(project: Project): Boolean {
    return true
  }
}
